reactive/kotlinx-coroutines-reactive/src/Await.kt

```diff
@@ -77,10 +80,21 @@ public suspend fun <T> Publisher<T>.awaitLast(): T = awaitOne(Mode.LAST)
  * @throws NoSuchElementException if publisher does not emit any value
  * @throws IllegalArgumentException if publisher emits more than one value
  */
-public suspend fun <T> Publisher<T>.awaitSingle(): T = awaitOne(Mode.SINGLE)
+public suspend fun <T> Publisher<T>.awaitSingle(): T = injectCoroutineContext().awaitOne(Mode.SINGLE)
 
 // ------------------------ private ------------------------
 
+// ContextInjector service is implemented in `kotlinx-coroutines-reactor` module only.
+// If `kotlinx-coroutines-reactor` module is not included, the list is empty.
+private val contextInjectors: List<ContextInjector> =
+    ServiceLoader.load(ContextInjector::class.java, ContextInjector::class.java.classLoader).toList()
+
+private suspend fun <T> Publisher<T>.injectCoroutineContext(): Publisher<T> {
```

Comment:
This function must have `context: CoroutineContext` as its parameter and it must not have `suspend` modifier. Note that all `awaitXxx` suspending functions remain ammendeable to tail-call optimization and do not have a state-machine. But you'd also move call to `injectCoroutineContext` to inside of `awaitOne` implementation and there you'd get coroutine context from the continuation via `cont.coroutineContext`.